subiquity/models/filesystem.py

```diff
@@ -646,3 +69,599 @@ def _dehumanize_size(size):
+    preserve = attr.ib(default=False)
+
+    _fs = attr.ib(default=None, repr=False) # Filesystem
+
+    @property
+    def available(self):
+        if self.flag == 'bios_grub':
+            return False
+        if self._fs is None:
+            return True
+        if self._fs._mount is None:
+            fs_obj = FilesystemModel.fs_by_name[self._fs.fstype]
+            return fs_obj.is_mounted
+        return False
+
+    @property
+    def path(self):
+        return "%s%s"%(self.device.path, self.number)
+
+
+@attr.s
+class Filesystem:
+    id = attr.ib(default=id_factory("fs"))
+    type = attr.ib(default="format")
+    fstype = attr.ib(default=None)
+    volume = attr.ib(default=None) # Partition or Disk
+    label = attr.ib(default=None)
+    uuid = attr.ib(default=None)
+    preserve = attr.ib(default=False)
+    _mount = attr.ib(default=None, repr=False) # Mount
+
+
+@attr.s
+class Mount:
+    id = attr.ib(default=id_factory("mount"))
+    type = attr.ib(default="mount")
+    device = attr.ib(default=None) # Filesystem
+    path = attr.ib(default=None)
+
+
+def align_up(size, block_size=1 << 20):
+    return (size + block_size - 1) & ~(block_size - 1)
+
+GPT_END_RESERVE = 1 << 20
+
+class FilesystemModel(object):
+
+    supported_filesystems = [
+        ('ext4', True, FS('ext4', True)),
+        ('xfs', True, FS('xfs', True)),
+        ('btrfs', True, FS('btrfs', True)),
+        ('---', False),
+        ('swap', True, FS('swap', False)),
+        #('bcache cache', True, FS('bcache cache', False)),
+        #('bcache store', True, FS('bcache store', False)),
+        ('---', False),
+        ('leave unformatted', True, FS(None, False)),
+    ]
+
+    fs_by_name = {}
+    for t in supported_filesystems:
+        if len(t) > 2:
+            fs = t[2]
+            if fs.label is not None:
+                fs_by_name[fs.label] = fs
+
+    def __init__(self, prober, opts):
+        self.prober = prober
+        self.opts = opts
+        self._available_disks = {} # keyed by path, eg /dev/sda
+        self.reset()
+
+    def reset(self):
+        self._disks = collections.OrderedDict() # only gets populated when something uses the disk
+        self._filesystems = []
+        self._partitions = []
+        self._mounts = []
+        for k, d in self._available_disks.items():
+            self._available_disks[k] = Disk.from_info(d._info)
+
+    def render(self):
+        r = []
+        for d in self._disks.values():
+            r.append(asdict(d))
+        for p in self._partitions:
+            r.append(asdict(p))
+        for f in self._filesystems:
+            r.append(asdict(f))
+        for m in self._mounts:
+            r.append(asdict(m))
+        return r
+
+    def _get_system_mounted_disks(self):
+        # This assumes a fairly vanilla setup. It won't list as
+        # mounted a disk that is only mounted via lvm, for example.
+        mounted_devs = []
+        with open('/proc/mounts') as pm:
```

Comment:
Er hm, I guess? This is linux though, it might well just be bytes. Although pyudev appears to assume sys.getfilesystemencoding() so I may as well too.